pkg/routeagent/controllers/route/route.go

```diff
@@ -24,30 +32,74 @@ type Controller struct {
 	clusterID       string
 	objectNamespace string
 
-	submarinerClientSet clientset.Interface
-	clustersSynced      cache.InformerSynced
-	endpointsSynced     cache.InformerSynced
+	submarinerClientSet    clientset.Interface
+	clientSet              *kubernetes.Clientset
+	clustersSynced         cache.InformerSynced
+	endpointsSynced        cache.InformerSynced
+	smRouteAgentPodsSynced cache.InformerSynced
 
 	clusterWorkqueue  workqueue.RateLimitingInterface
 	endpointWorkqueue workqueue.RateLimitingInterface
+	podWorkqueue      workqueue.RateLimitingInterface
 
-	gw      net.IP
-	subnets []string
+	gatewayNodeIP    net.IP
+	localClusterCidr []string
+	localServiceCidr []string
+	remoteSubnets    *util.StringSet
 
-	link *net.Interface
+	vxlanDevice *vxLanIface
```

Comment:
`vxlanDevice` is shared between the _Pod_ and _Endpoint_ processing and thus can be accessed by multiple threads so needs to be synchronized or use the same workqueue goroutine for both _Pod_ and _Endpoint_.  I think the latter would be safest to avoid any potential concurrency issues (also wrt `remoteVTEPs` and `isGatewayNode`).

Reply:
One potential issue is that the `vxlanDevice` variable is not synchronized so it's possible, albeit unlikely,  that the thread processing pods never sees the value for `vxlanDevice` written by the thread processing endpoints due to CPU caching. To be completely safe, all accesses of `vxlanDevice` need to be protected by a mutex or `vxlanDevice` needs to be an atomic reference.  Similarly with `isGatewayNode`. 

Another possible issue is that you set `isGatewayNode` before  `vxlanDevice` however w/o synchronization, it's possible that if the pod thread ran concurrently, it may observe the update to `vxlanDevice` but not `isGatewayNode` which would lead to incorrect state (ie it may not call `AddFDB` when it should).  There may be other concurrency issues lurking. It may be that things would eventually converge properly due to the 30 sec resync period but it's very difficult/impossible to think of and test all such scenarios and ensure correctness. So I think it's safest to take concurrency out of the equation altogether by using the same workqueue for both pods and endpoints rather than using shared memory.

Reply:
Given that we can't enqueue _removes_ as mentioned in a prior comment, it wouldn't buy us anything to share the same workqueue as we need shared memory anyway. However I still think we should synchronize access to `vxlanDevice` and `isGatewayNode` atomically to be safe.

Reply:
I'm not sure why this conversation was marked as resolved. We still need to address synchronization of `vxlanDevice` and `isGatewayNode` as discussed above.

Reply:
Oops, it was accidentally marked as resolved.

Reply:
Yeah @sridhargaddam remember the thread about memory synchronization between CPU cores we had. If we don't mutex, we have no guarantee that the vxlanDevice will be read by other threads than the one which wrote it, although, caches are small and it's very likely, what when it hits a processor with a bigger cache?